examples/fluids/qfunctions/densitycurrent.h

```diff
@@ -148,4 +148,81 @@ CEED_QFUNCTION(ICsDC)(void *ctx, CeedInt Q,
   return 0;
 }
 
+// *****************************************************************************
+// This helper function provides support for the exact, time-dependent solution
+//   (currently not implemented) and IC formulation for density current
+//   with primitive variables
+// *****************************************************************************
+CEED_QFUNCTION_HELPER int Exact_DC_Prim(CeedInt dim, CeedScalar time,
+                                        const CeedScalar X[], CeedInt Nf, CeedScalar q[],
+                                        void *ctx) {
+  // Context
+  const SetupContext context = (SetupContext)ctx;
+  const CeedScalar theta0   = context->theta0;
+  const CeedScalar thetaC   = context->thetaC;
+  const CeedScalar N        = context->N;
+  const CeedScalar cp       = context->cp;
+  const CeedScalar *g_vec   = context->g;
+  const CeedScalar rc       = context->rc;
+  const CeedScalar *center  = context->center;
+  const CeedScalar *dc_axis = context->dc_axis;
+  const CeedScalar g = -g_vec[2];
+
+  // Setup
+  // -- Coordinates
+  const CeedScalar x = X[0];
+  const CeedScalar y = X[1];
+  const CeedScalar z = X[2];
+
+  // -- Potential temperature, density current
+  CeedScalar rr[3] = {x - center[0], y - center[1], z - center[2]};
+  // (I - q q^T) r: distance from dc_axis (or from center if dc_axis is the zero vector)
+  for (CeedInt i=0; i<3; i++)
+    rr[i] -= dc_axis[i] *
+             (dc_axis[0]*rr[0] + dc_axis[1]*rr[1] + dc_axis[2]*rr[2]);
+  const CeedScalar r = sqrt(rr[0]*rr[0] + rr[1]*rr[1] + rr[2]*rr[2]);
+  const CeedScalar delta_theta = r <= rc ? thetaC*(1. + cos(M_PI*r/rc))/2. : 0.;
+  const CeedScalar theta = theta0*exp(N*N*z/g) + delta_theta;
+
+  // -- Exner pressure, hydrostatic balance
+  const CeedScalar Pi = 1. + g*g*(exp(-N*N*z/g) - 1.) / (cp*theta0*N*N);
+
+  // Initial Conditions
+  q[0] = Pi;
+  q[1] = 0.0;
+  q[2] = 0.0;
+  q[3] = 0.0;
+  q[4] = theta;
```

Comment:
Here also, could we have a helper that returns `State` and then `ICsDC_Cons` and `ICsDC_Prim` would just unpack the variables it needs?